File: pkg/util/test/events_test.go
// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package test_test

import (
	"context"
	"testing"
	"time"

	"github.com/smartystreets/assertions"
	"go.thethings.network/lorawan-stack/pkg/events"
	"go.thethings.network/lorawan-stack/pkg/ttnpb"
	. "go.thethings.network/lorawan-stack/pkg/util/test"
	"go.thethings.network/lorawan-stack/pkg/util/test/assertions/should"
)

func TestSetDefaultEventPubSub(t *testing.T) {
	a := assertions.New(t)

	ctx := ContextWithT(Context(), t)
	ctx, cancel := context.WithTimeout(ctx, (1<<5)*Delay)
	defer cancel()

	testEvent1 := events.Define("test-set-default-event-pub-sub-1", "test-event-1")(ctx, ttnpb.ApplicationIdentifiers{ApplicationID: "test-app"}, nil)
	testEvent2 := events.Define("test-set-default-event-pub-sub-2", "test-event-2")(ctx, ttnpb.GatewayIdentifiers{GatewayID: "test-gtw"}, nil)

	events.Publish(testEvent1)
	events.Publish(testEvent2)

	publishCh1 := make(chan EventPubSubPublishRequest)
	undo1 := SetDefaultEventsPubSub(&MockEventPubSub{
		PublishFunc: MakeEventPubSubPublishChFunc(publishCh1),
	})
	go events.Publish(testEvent1)
	if !AssertEventPubSubPublishRequest(ctx, publishCh1, func(ev events.Event) bool {
		return a.So(ev, should.Equal, testEvent1)
	}) {
		t.FailNow()
	}
	close(publishCh1)

	publishCh2 := make(chan EventPubSubPublishRequest)
	time.AfterFunc(Delay, undo1)
	undo2 := SetDefaultEventsPubSub(&MockEventPubSub{
		PublishFunc: MakeEventPubSubPublishChFunc(publishCh2),
	})
	go events.Publish(testEvent2)
	if !AssertEventPubSubPublishRequest(ctx, publishCh2, func(ev events.Event) bool {
		return a.So(ev, should.Equal, testEvent2)
	}) {
		t.FailNow()
	}
	close(publishCh2)
	undo2()

	events.Publish(testEvent1)
	events.Publish(testEvent2)
}

func TestCollectEvents(t *testing.T) {
	ctx := Context()
	testEvent1 := events.Define("test-collect-events-1", "test-event-1")(ctx, ttnpb.ApplicationIdentifiers{ApplicationID: "test-app"}, nil)
	testEvent2 := events.Define("test-collect-events-2", "test-event-2")(ctx, ttnpb.GatewayIdentifiers{GatewayID: "test-gtw"}, nil)
	assertions.New(t).So(CollectEvents(func() {
		events.Publish(testEvent1)
		events.Publish(testEvent1)
		events.Publish(testEvent2)
	}), should.Resemble, []events.Event{
		testEvent1,
		testEvent1,
		testEvent2,
	})
}

File: pkg/webui/console/store/reducers/tests/events_test.js
// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import createReducer from '../events'
import { getEventMessageSuccess } from '../../actions/events'

describe('Events reducer', () => {
  const testId = 'test'
  const reducer = createReducer(testId)
  const successActionCreator = getEventMessageSuccess(testId)

  describe('when adding new events', () => {
    it('keeps events sorted by `time`', () => {
      const testId = 'test'
      const testTime = '2019-03-28T13:18:13.376022Z'
      const testDate = new Date(testTime)

      const initialState = reducer(undefined, {})

      const event1 = { time: testTime }
      let newState = reducer(initialState, successActionCreator(testId, event1))

      expect(newState[testId].events).toHaveLength(1)
      expect(newState[testId].events[0]).toEqual(event1)

      const event2 = { time: new Date(testDate.getTime() + 1000).toISOString() }
      newState = reducer(newState, successActionCreator(testId, event2))

      expect(newState[testId].events).toHaveLength(2)
      expect(newState[testId].events[0]).toEqual(event2)
      expect(newState[testId].events[1]).toEqual(event1)

      const event3 = { time: new Date(testDate.getTime() - 1000).toISOString() }
      newState = reducer(newState, successActionCreator(testId, event3))

      expect(newState[testId].events).toHaveLength(3)
      expect(newState[testId].events[0]).toEqual(event2)
      expect(newState[testId].events[1]).toEqual(event1)
      expect(newState[testId].events[2]).toEqual(event3)

      const event4 = { time: new Date(testDate.getTime() + 2000).toISOString() }
      newState = reducer(newState, successActionCreator(testId, event4))

      expect(newState[testId].events).toHaveLength(4)
      expect(newState[testId].events[0]).toEqual(event4)
      expect(newState[testId].events[1]).toEqual(event2)
      expect(newState[testId].events[2]).toEqual(event1)
      expect(newState[testId].events[3]).toEqual(event3)

      const event5 = { time: new Date(testDate.getTime()).toISOString() }
      newState = reducer(newState, successActionCreator(testId, event5))

      expect(newState[testId].events).toHaveLength(5)
      expect(newState[testId].events[0]).toEqual(event4)
      expect(newState[testId].events[1]).toEqual(event2)
      expect(newState[testId].events[2]).toEqual(event5)
      expect(newState[testId].events[3]).toEqual(event1)
      expect(newState[testId].events[4]).toEqual(event3)
    })
  })
})
